lib/vagrant/action/builtin/cloud_init_setup.rb

```diff
@@ -0,0 +1,130 @@
+          end
+
+          # Continue On
+          @app.call(env)
+        end
+
+        # @param [Vagrant::Machine] machine
+        # @param [Vagrant::Environment] env
+        # @param [Array<#VagrantPlugins::Kernel_V2::VagrantConfigCloudInit>] user_data_cfgs
+        # @return [MIME::Text] user_data
+        def setup_user_data(machine, env, user_data_cfgs)
+          machine.ui.info(I18n.t("vagrant.actions.vm.cloud_init_user_data_setup"))
+
+          text_cfgs = []
+          user_data_cfgs.each do |cfg|
+            text_cfgs << read_text_cfg(machine, cfg)
+          end
+
+          user_data = generate_cfg_msg(machine, text_cfgs)
+          user_data
+        end
+
+        # Reads an individual cloud_init config and stores its contents and the
+        # content_type as a MIME text
+        #
+        # @param [Vagrant::Machine] machine
+        # @param [VagrantPlugins::Kernel_V2::VagrantConfigCloudInit] cfg
+        # @return [MIME::Text] text_msg
+        def read_text_cfg(machine, cfg)
+          if cfg.path
+            text = File.read(Pathname.new(cfg.path).expand_path(machine.env.root_path))
+          else
+            text = cfg.inline
+          end
+
+          # Note: content_type must remove the leading `text/` because
+          # the MIME::Text initializer hardcodes `text/` already to the type.
+          # We assume content_type is correct due to the validation step
+          # in VagrantConfigCloudInit.
+          content_type = cfg.content_type.split('/')
+          text_msg = MIME::Text.new(text, content_type[1])
+
+          text_msg
+        end
+
+        # Combines all known cloud_init configs into a multipart mixed MIME text
+        # message
+        #
+        # @param [Vagrant::Machine] machine
+        # @param [Array<MIME::Text>] text_msg - One or more text configs
+        # @return [MIME::Multipart::Mixed] msg
+        def generate_cfg_msg(machine, text_cfgs)
+          msg = MIME::Multipart::Mixed.new
+
+          text_cfgs.each do |c|
+            msg.add(c)
+          end
+
+          msg
+        end
+
+        # Writes the contents of the guests cloud_init config to a tmp
+        # dir and passes that source directory along to the host cap to be
+        # written to an iso
+        #
+        # @param [Vagrant::Machine] machine
+        # @param [MIME::Multipart::Mixed] user_data
+        # @param [Hash] meta_data
+        def write_cfg_iso(machine, env, user_data, meta_data)
+          iso_path = nil
+
+          if env[:env].host.capability?(:create_iso)
+            begin
+              source_dir = Pathname.new(Dir.mktmpdir(TEMP_PREFIX))
+              File.open("#{source_dir}/user-data", 'w') { |file| file.write(user_data.to_s) }
+
+              File.open("#{source_dir}/meta-data", 'w') { |file| file.write(meta_data.to_s) }
+
+              iso_path = env[:env].host.capability(:create_iso, env[:env],
+                                                   source_dir, volume_id: "cidata")

```

Comment:
the current pr for this host cap accepts an opts hash like `extra_opts: {:volume_id => "cidata"}`. I'm happy to change that pr if it makes more sense that way.

Reply:
@soapy1 I see! Hmm, yeah, I think it would be cleaner if the `create_iso` cap defined the extra options with something like:

```ruby
def create_iso(env, dir, **extra_opts)
 # stuff
end
```

Then it can work like how it's defined above where the final param is a hash of options rather than needing to pass a key that has its value as a hash.

Reply:
Also I think code wise nothing should change, you should be able to access the keys inside `extra_opts` the same way that you are now (like `extra_opts[:key]`), so the only thing that changes is the parameter!